### PR TITLE
Implement new stats blocks

### PR DESCRIPTION
**Pull request by daki7711**
First attempt to change the min max block in the list category. Note that at the moment I implemented finding min and max using sorting, will optimize it.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,22 @@
+.gitignore
```
Comment (ewpatton) [inline .gitignore:1]: Please remove the .gitignore file from the branch.

File: appinventor/blocklyeditor/src/blocks/lists.js [hunk 2]
```diff
@@ -790,6 +790,61 @@ Blockly.Blocks['lists_minimum_number'] = {
   category : 'Lists',
   helpUrl : Blockly.Msg.LANG_LISTS_MIN_NUMBER_HELPURL,
   init : function() {
+      this.setColour(Blockly.LIST_CATEGORY_HUE);
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/lists.js:793]: There's too much indentation here. This line and the ones following should only be +2.

File: appinventor/blocklyeditor/src/blocks/lists.js [hunk 3]
```diff
@@ -790,6 +790,61 @@ Blockly.Blocks['lists_minimum_number'] = {
...
+      var result = Blockly.LexicalVariable.freeVariables(this.getInputTargetBlock(this.bodyInputName));
+      result.subtract(new Blockly.NameSet(this.declaredNames()));
+      if (result.isEmpty()) {
+        return result;
+      } else {
+        throw "Violation of invariant: procedure declaration has nonempty free variables: " + result.toString();
+      }
+    },
+    typeblock: [{ translatedName: Blockly.Msg.LANG_LISTS_MIN_NUMBER_INPUT_COLLAPSED_TEXT }]
+  /*init : function() {
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/lists.js:847]: Remove the commented out code.

File: appinventor/blocklyeditor/src/blocks/lists.js [hunk 4]
```diff
@@ -798,21 +853,77 @@ Blockly.Blocks['lists_minimum_number'] = {
...
   typeblock: [{ translatedName: Blockly.Msg.LANG_LISTS_MIN_NUMBER_INPUT_MIN }]
+*/
 };
 
 Blockly.Blocks['lists_maximum_number'] = {
   // Maximum value in the list
   category : 'Lists',
   helpUrl : Blockly.Msg.LANG_LISTS_MAX_NUMBER_HELPURL,
   init : function() {
+        this.setColour(Blockly.LIST_CATEGORY_HUE);
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/lists.js:864]: Ditto indentation

File: appinventor/blocklyeditor/src/blocks/lists.js [hunk 5]
```diff
@@ -798,21 +853,77 @@ Blockly.Blocks['lists_minimum_number'] = {
...
+        var result = Blockly.LexicalVariable.freeVariables(this.getInputTargetBlock(this.bodyInputName));
+        result.subtract(new Blockly.NameSet(this.declaredNames()));
+        if (result.isEmpty()) {
+          return result;
+        } else {
+          throw "Violation of invariant: procedure declaration has nonempty free variables: " + result.toString();
+        }
+      },
+      typeblock: [{ translatedName: Blockly.Msg.LANG_LISTS_MAX_NUMBER_COLLAPSED_TEXT }]
+  /*init : function() {
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/lists.js:918]: Ditto removing code

File: appinventor/blocklyeditor/src/blocks/lists.js [hunk 6]
```diff
@@ -798,21 +853,77 @@ Blockly.Blocks['lists_minimum_number'] = {
...
+        // Should return the empty set: something is wrong if it doesn't!
+        var result = Blockly.LexicalVariable.freeVariables(this.getInputTargetBlock(this.bodyInputName));
+        result.subtract(new Blockly.NameSet(this.declaredNames()));
+        if (result.isEmpty()) {
+          return result;
+        } else {
+          throw "Violation of invariant: procedure declaration has nonempty free variables: " + result.toString();
+        }
+      },
+      typeblock: [{ translatedName: Blockly.Msg.LANG_LISTS_MAX_NUMBER_COLLAPSED_TEXT }]
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/lists.js:917]: What is the significance of using the collapsed text message here?

File: appinventor/blocklyeditor/src/blocks/lists.js [hunk 7]
```diff
@@ -798,21 +853,77 @@ Blockly.Blocks['lists_minimum_number'] = {
...
+        }
+      },
+      freeVariables: function() { // return the free lexical variables of this block
+        // Should return the empty set: something is wrong if it doesn't!
+        var result = Blockly.LexicalVariable.freeVariables(this.getInputTargetBlock(this.bodyInputName));
+        result.subtract(new Blockly.NameSet(this.declaredNames()));
+        if (result.isEmpty()) {
+          return result;
+        } else {
+          throw "Violation of invariant: procedure declaration has nonempty free variables: " + result.toString();
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/lists.js:914]: This throw message should probably be wrapped in an `Error()`.

File: appinventor/blocklyeditor/src/blocks/lists.js [hunk 8]
```diff
@@ -790,6 +790,61 @@ Blockly.Blocks['lists_minimum_number'] = {
...
+      }
+    },
+    freeVariables: function() { // return the free lexical variables of this block
+      // Should return the empty set: something is wrong if it doesn't!
+      var result = Blockly.LexicalVariable.freeVariables(this.getInputTargetBlock(this.bodyInputName));
+      result.subtract(new Blockly.NameSet(this.declaredNames()));
+      if (result.isEmpty()) {
+        return result;
+      } else {
+        throw "Violation of invariant: procedure declaration has nonempty free variables: " + result.toString();
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/lists.js:843]: This error message should be wrapped in a call to `Error()`.

File: appinventor/blocklyeditor/src/blocks/lists.js [hunk 9]
```diff
@@ -790,6 +790,61 @@ Blockly.Blocks['lists_minimum_number'] = {
...
+      }
+    },
+    typeblock: [{ translatedName: Blockly.Msg.LANG_LISTS_MIN_NUMBER_INPUT_COLLAPSED_TEXT }]
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/lists.js:846]: Why are we using the collapsed text for typeblock?

File: appinventor/blocklyeditor/src/field_lexical_variable.js [hunk 10]
```diff
@@ -323,8 +323,20 @@ Blockly.FieldLexicalVariable.getLexicalNamesInScope = function (block) {
             rememberName(loopName2, loopNames, Blockly.loopParameterPrefix);
           } else if ( (parent.type === "lists_sort_key")
                   && (parent.getInputTargetBlock('KEY') == child)) {// Only DO is in scope, not other inputs!
-           var loopName = parent.getFieldValue('VAR');
-           rememberName(loopName, loopNames, Blockly.loopParameterPrefix);
+            var loopName = parent.getFieldValue('VAR');
+            rememberName(loopName, loopNames, Blockly.loopParameterPrefix);
+          } else if ( (parent.type === "lists_minimum_number")
+                  && (parent.getInputTargetBlock('COMPARE') == child)) {// Only DO is in scope, not other inputs!
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/field_lexical_variable.js:329]: Code style: Indentation.

Also, I'm not sure the comment on this line is correct.

File: appinventor/blocklyeditor/src/field_lexical_variable.js [hunk 11]
```diff
@@ -323,8 +323,20 @@ Blockly.FieldLexicalVariable.getLexicalNamesInScope = function (block) {
...
+            var loopName1 = parent.getFieldValue('VAR1');
+            var loopName2 = parent.getFieldValue('VAR2');
+            rememberName(loopName1, loopNames, Blockly.loopParameterPrefix);
+            rememberName(loopName2, loopNames, Blockly.loopParameterPrefix);
+          } else if ( (parent.type === "lists_maximum_number")
+                 && (parent.getInputTargetBlock('COMPARE') == child)) {// Only DO is in scope, not other inputs!
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/field_lexical_variable.js:335]: Ditto.

File: appinventor/docs/html/reference/components/extension.html [hunk 12]
```diff
@@ -0,0 +1,301 @@
+<!DOCTYPE html>
```
Comment (ewpatton) [inline appinventor/docs/html/reference/components/extension.html:1]: Please remove this file.

File: appinventor/docs/markdown/reference/components/extension.md [hunk 13]
```diff
@@ -0,0 +1,72 @@
+---
```
Comment (ewpatton) [inline appinventor/docs/markdown/reference/components/extension.md:1]: Please remove this file.

File: appinventor/misc/emulator-support/aiStarter.py [hunk 14]
```diff
@@ -14,7 +14,7 @@
 PLATDIR = os.path.abspath(os.path.dirname(sys.argv[0]))
 
 # Path to executables
-ADB = os.path.join(PLATDIR, 'from-Android-SDK', 'platform-tools', 'adb')
+ADB = "/Users/davidkim/Library/Android/sdk/platform-tools/adb"
```
Comment (ewpatton) [inline appinventor/misc/emulator-support/aiStarter.py:17]: Reset this unrelated change.

File: appinventor/docs/html/reference/components/userinterface.html [hunk 15]
```diff
@@ -172,9 +172,10 @@ <h3 id="Button-Properties">Properties</h3>
  Some fonts do not support italic.</dd>
   <dt id="Button.FontSize" class="number"><em>FontSize</em></dt>
   <dd>Specifies the text font size of the <code class="highlighter-rouge">Button</code>, measured in sp(scale-independent pixels).</dd>
-  <dt id="Button.FontTypeface" class="number do"><em>FontTypeface</em></dt>
+  <dt id="Button.FontTypeface" class="text do"><em>FontTypeface</em></dt>
```
Comment (ewpatton) [inline appinventor/docs/html/reference/components/userinterface.html:175]: I would reset this file to be the same as on ucr.

File: appinventor/docs/html/reference/components/social.html [hunk 16]
```diff
@@ -189,9 +189,10 @@ <h3 id="ContactPicker-Properties">Properties</h3>
  Some fonts do not support italic.</dd>
   <dt id="ContactPicker.FontSize" class="number"><em>FontSize</em></dt>
   <dd>Specifies the text font size of the <code class="highlighter-rouge">ContactPicker</code>, measured in sp(scale-independent pixels).</dd>
-  <dt id="ContactPicker.FontTypeface" class="number do"><em>FontTypeface</em></dt>
+  <dt id="ContactPicker.FontTypeface" class="text do"><em>FontTypeface</em></dt>
```
Comment (ewpatton) [inline appinventor/docs/html/reference/components/social.html:192]: I would reset this file to match ucr.

File: appinventor/docs/html/reference/components/media.html [hunk 17]
```diff
@@ -227,9 +227,10 @@ <h3 id="ImagePicker-Properties">Properties</h3>
  Some fonts do not support italic.</dd>
   <dt id="ImagePicker.FontSize" class="number"><em>FontSize</em></dt>
   <dd>Specifies the text font size of the <code class="highlighter-rouge">ImagePicker</code>, measured in sp(scale-independent pixels).</dd>
-  <dt id="ImagePicker.FontTypeface" class="number do"><em>FontTypeface</em></dt>
+  <dt id="ImagePicker.FontTypeface" class="text do"><em>FontTypeface</em></dt>
```
Comment (ewpatton) [inline appinventor/docs/html/reference/components/media.html:230]: I would reset this file to match ucr.

File: appinventor/blocklyeditor/src/generators/yail/lists.js [hunk 18]
```diff
@@ -422,28 +422,34 @@ Blockly.Yail['lists_sort_key'] = function() {
...
-  var argument0 = Blockly.Yail.valueToCode(this, 'LIST', Blockly.Yail.ORDER_NONE) || Blockly.Yail.emptyListCode;
-  var code = Blockly.Yail.YAIL_CALL_YAIL_PRIMITIVE + "yail-list-minimum-number" + Blockly.Yail.YAIL_SPACER;
-  code = code + Blockly.Yail.YAIL_OPEN_COMBINATION + Blockly.Yail.YAIL_LIST_CONSTRUCTOR + Blockly.Yail.YAIL_SPACER;
-  code = code + argument0;
-  code = code + Blockly.Yail.YAIL_SPACER + Blockly.Yail.YAIL_CLOSE_COMBINATION;
-  code = code + Blockly.Yail.YAIL_SPACER + Blockly.Yail.YAIL_QUOTE + Blockly.Yail.YAIL_OPEN_COMBINATION;
-  code = code + "list" + Blockly.Yail.YAIL_CLOSE_COMBINATION + Blockly.Yail.YAIL_SPACER;
-  code = code + Blockly.Yail.YAIL_DOUBLE_QUOTE + "minimum number of list" + Blockly.Yail.YAIL_DOUBLE_QUOTE + Blockly.Yail.YAIL_CLOSE_COMBINATION;
-  return [ code, Blockly.Yail.ORDER_ATOMIC ];
+  var emptyListCode = Blockly.Yail.YAIL_CALL_YAIL_PRIMITIVE + "make-yail-list" + Blockly.Yail.YAIL_SPACER;
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/generators/yail/lists.js:425]: Fix the indentation in this file.

File: appinventor/blocklyeditor/src/generators/yail/lists.js [hunk 19]
```diff
@@ -422,28 +422,34 @@ Blockly.Yail['lists_sort_key'] = function() {
...
+  	emptyListCode += Blockly.Yail.YAIL_OPEN_COMBINATION + Blockly.Yail.YAIL_LIST_CONSTRUCTOR + Blockly.Yail.YAIL_SPACER;
+  	emptyListCode += Blockly.Yail.YAIL_CLOSE_COMBINATION + Blockly.Yail.YAIL_SPACER + Blockly.Yail.YAIL_QUOTE + Blockly.Yail.YAIL_OPEN_COMBINATION;
+  	emptyListCode += Blockly.Yail.YAIL_CLOSE_COMBINATION;
+  	emptyListCode += Blockly.Yail.YAIL_SPACER + Blockly.Yail.YAIL_DOUBLE_QUOTE + "minimum number of list" + Blockly.Yail.YAIL_DOUBLE_QUOTE + Blockly.Yail.YAIL_CLOSE_COMBINATION;
+  	var loopIndexName1 = Blockly.Yail.YAIL_LOCAL_VAR_TAG + this.getFieldValue('VAR1');
+  	var loopIndexName2 = Blockly.Yail.YAIL_LOCAL_VAR_TAG + this.getFieldValue('VAR2');
+  	var listCode = Blockly.Yail.valueToCode(this, 'LIST', Blockly.Yail.ORDER_NONE) || emptyListCode;
+  	var bodyCode = Blockly.Yail.valueToCode(this, 'COMPARE', Blockly.Yail.ORDER_NONE) ||  Blockly.Yail.YAIL_FALSE;
+  	var code = "(mincomparator_nondest " + loopIndexName1 + Blockly.Yail.YAIL_SPACER + loopIndexName2 + Blockly.Yail.YAIL_SPACER + bodyCode + Blockly.Yail.YAIL_SPACER
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/generators/yail/lists.js:434]: I'd prefer to use hyphens rather than underscores in the Scheme names.

File: appinventor/blocklyeditor/src/generators/yail/lists.js [hunk 20]
```diff
@@ -422,28 +422,34 @@ Blockly.Yail['lists_sort_key'] = function() {
...
+  var emptyListCode = Blockly.Yail.YAIL_CALL_YAIL_PRIMITIVE + "make-yail-list" + Blockly.Yail.YAIL_SPACER;
+  	emptyListCode += Blockly.Yail.YAIL_OPEN_COMBINATION + Blockly.Yail.YAIL_LIST_CONSTRUCTOR + Blockly.Yail.YAIL_SPACER;
+  	emptyListCode += Blockly.Yail.YAIL_CLOSE_COMBINATION + Blockly.Yail.YAIL_SPACER + Blockly.Yail.YAIL_QUOTE + Blockly.Yail.YAIL_OPEN_COMBINATION;
+  	emptyListCode += Blockly.Yail.YAIL_CLOSE_COMBINATION;
+  	emptyListCode += Blockly.Yail.YAIL_SPACER + Blockly.Yail.YAIL_DOUBLE_QUOTE + "maximum number of list" + Blockly.Yail.YAIL_DOUBLE_QUOTE + Blockly.Yail.YAIL_CLOSE_COMBINATION;
+  	var loopIndexName1 = Blockly.Yail.YAIL_LOCAL_VAR_TAG + this.getFieldValue('VAR1');
+  	var loopIndexName2 = Blockly.Yail.YAIL_LOCAL_VAR_TAG + this.getFieldValue('VAR2');
+  	var listCode = Blockly.Yail.valueToCode(this, 'LIST', Blockly.Yail.ORDER_NONE) || emptyListCode;
+  	var bodyCode = Blockly.Yail.valueToCode(this, 'COMPARE', Blockly.Yail.ORDER_NONE) ||  Blockly.Yail.YAIL_FALSE;
+  	var code = "(maxcomparator_nondest " + loopIndexName1 + Blockly.Yail.YAIL_SPACER + loopIndexName2 + Blockly.Yail.YAIL_SPACER + bodyCode + Blockly.Yail.YAIL_SPACER
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/generators/yail/lists.js:450]: Ditto.

**Review by ewpatton — CHANGES_REQUESTED** [review]

File: appinventor/blocklyeditor/src/blocks/lists.js (hunk 8 above)
Comment (daki7711) [reply appinventor/blocklyeditor/src/blocks/lists.js:843]: So would it be Error("Violation of invariant: procedure declaration has nonempty free variables: " + result.toString());? Cause this is how it is implemented with other blocks

File: appinventor/blocklyeditor/src/blocks/lists.js (hunk 9 above)
Comment (daki7711) [reply appinventor/blocklyeditor/src/blocks/lists.js:846]: Similar to above, the other blocks seem to use typeblock, do you think it should be something else?

File: appinventor/blocklyeditor/src/msg/ai_blockly/messages.json [hunk 21]
```diff
@@ -575,10 +578,22 @@
 
 	"Blockly.Msg.LANG_LISTS_MIN_NUMBER_HELPURL": "http://appinventor.mit.edu/explore/ai2/support/blocks/lists#minnumber",
 	"Blockly.Msg.LANG_LISTS_MIN_NUMBER_INPUT_MIN": "minimum number in the list",
+	"Blockly.Msg.LANG_LISTS_MIN_NUMBER_INPUT_COMPARATOR": "where ",
+	"Blockly.Msg.LANG_LISTS_MIN_NUMBER_INPUT_VAR1": "item1",
+	"Blockly.Msg.LANG_LISTS_MIN_NUMBER_INPUT_AND": "precedes",
+	"Blockly.Msg.LANG_LISTS_MIN_NUMBER_INPUT_VAR2": "item2",
+	"Blockly.Msg.LANG_LISTS_MIN_NUMBER_INPUT_COMPARATOR2": "if",
+	"Blockly.Msg.LANG_LISTS_MIN_NUMBER_INPUT_COLLAPSED_TEXT": "minimum using comparator",
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/msg/ai_blockly/messages.json:586]: `Blockly.Msg.LANG_LISTS_MIN_NUMBER_TYPEBLOCK`

File: appinventor/blocklyeditor/src/msg/ai_blockly/messages.json [hunk 22]
```diff
@@ -575,10 +578,22 @@
...
 	"Blockly.Msg.LANG_LISTS_MIN_NUMBER_TOOLTIP": "Returns minimum number in the list",
 
 	"Blockly.Msg.LANG_LISTS_MAX_NUMBER_HELPURL": "http://appinventor.mit.edu/explore/ai2/support/blocks/lists#maxnumber",
 	"Blockly.Msg.LANG_LISTS_MAX_NUMBER_INPUT_MAX": "maximum number in the list",
+	"Blockly.Msg.LANG_LISTS_MAX_NUMBER_INPUT_COMPARATOR": "where ",
+	"Blockly.Msg.LANG_LISTS_MAX_NUMBER_INPUT_VAR1": "item1",
+	"Blockly.Msg.LANG_LISTS_MAX_NUMBER_INPUT_AND": "precedes",
+	"Blockly.Msg.LANG_LISTS_MAX_NUMBER_INPUT_VAR2": "item2",
+	"Blockly.Msg.LANG_LISTS_MAX_NUMBER_INPUT_COMPARATOR2": "if",
+	"Blockly.Msg.LANG_LISTS_MAX_NUMBER_COLLAPSED_TEXT": "maximum using comparator",
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/msg/ai_blockly/messages.json:596]: `Blockly.Msg.LANG_LISTS_MAX_NUMBER_TYPEBLOCK`
